### PR TITLE
ARROW-7379: [C++] Introduce SchemaBuilder companion class and Field::IsCompatibleWith

### PR DESCRIPTION
**Pull request by fsaintjacques**
The SchemaBuilder class is introduced to merge/unify schemas in a controlled fashion.

### CODE REVIEW

**Comment by fsaintjacques** [comment]
Please comment on naming of methods, namely I hesitated for `Field::WithField` vs `Field::MergeWith`.

File: cpp/src/arrow/type.h [hunk 1]
```diff
@@ -1546,6 +1581,97 @@ std::shared_ptr<Schema> schema(
...
+ public:
+  // Indicate how field conflict(s) should be resolved when building a schema. A
+  // conflict arise when a field is added to the builder and one or more field(s)
+  // with the same name already exists.
+  enum ConflictPolicy {
+    // Ignore the conflict and append the field. This is the default behavior of the
+    // Schema constructor and the `arrow::schema` factory function.
+    APPEND = 0,
+    // Replace the existing field with the newer one.
+    REPLACE,
```
Comment (fsaintjacques) [inline cpp/src/arrow/type.h]: I'm tempted to change this with `KEEP_OURS`, `KEEP_THEIRS` like in git merge.

File: cpp/src/arrow/type.h [hunk 2]
```diff
@@ -358,11 +359,41 @@ class ARROW_EXPORT Field : public detail::Fingerprintable {
...
+  /// The two fields must be compatible, i.e:
+  ///   - have the same name
+  ///   - have the same type, or one or both are NullType
+  ///
+  /// Nullability will be promoted to the looser option (nullable if one is not
+  /// nullable). This method does _not_ accommodate type conversion minus
+  /// nullability. This includes integer widening, promotion from integer to
+  /// float, or conversion to or from boolean.
+  ///
+  /// The metadata of the current field is preserved.
```
Comment (bkietz) [inline cpp/src/arrow/type.h]:
```suggestion
  /// The metadata of the current field is preserved; the metadata of other is discarded.
```

File: cpp/src/arrow/type.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/type.h]: I prefer `REPLACE` and `IGNORE`

File: cpp/src/arrow/type.h [hunk 3]
```diff
@@ -1546,6 +1581,97 @@ std::shared_ptr<Schema> schema(
 
 /// @}
 
+/// \brief Convenience class to incrementally construct/merge schemas.
+///
+/// This class amortizes the cost of validating field name conflicts by
+/// maintaining the mapping. The caller also controls the conflict resolution
+/// scheme.
+class SchemaBuilder {
```
Comment (bkietz) [inline cpp/src/arrow/type.h]: This should be internal

**Review by bkietz — COMMENTED** [review]
Some quick comments
Maybe it's time we moved schema to a separate header file?

File: cpp/src/arrow/type_test.cc [hunk 4]
```diff
@@ -134,14 +189,40 @@ TEST(TestField, Equals) {
...
+  Field f0_with_meta2("f0", int32(), true, meta2);
+  Field f0_with_meta3("f0", int32(), true, meta3);
+  Field other("other", int64());
+
+  AssertFieldsCompatible(f0, f0_other);
+  AssertFieldsCompatible(f0, f0_with_meta1);
+  AssertFieldsCompatible(f0, f0_nn);
+  AssertFieldsCompatible(f0, f0_nt);
+  AssertFieldsCompatible(f0_nt, f0_with_meta1);
+  AssertFieldsCompatible(f0_with_meta1, f0_with_meta2);
```
Comment (pitrou) [inline cpp/src/arrow/type_test.cc]: Is it desirable? We can keep this as a first approach but we may need to make it later tunable.

File: cpp/src/arrow/type.h [hunk 5]
```diff
@@ -1546,6 +1581,97 @@ std::shared_ptr<Schema> schema(
...
+  ConflictPolicy policy() const { return policy_; }
+
+  /// \brief Set the conflict resolution method.
+  void SetPolicy(ConflictPolicy resolution) { policy_ = resolution; }
+
+  /// \brief Add a field to the constructed schema.
+  ///
+  /// \param[in] field to add to the constructed Schema.
+  /// \return A failure if encountered.
+  Status WithField(const std::shared_ptr<Field>& field);
```
Comment (pitrou) [inline cpp/src/arrow/type.h]: Call this `AddField`, same for other methods below.

File: cpp/src/arrow/type.h [hunk 6]
```diff
@@ -1546,6 +1581,97 @@ std::shared_ptr<Schema> schema(
...
+  /// \return The first failure encountered, if any.
+  Status WithSchema(const std::shared_ptr<Schema>& schema);
+
+  /// \brief Add fields of multiple Schemas to the constructed Schema.
+  ///
+  /// \param[in] schemas to take fields to add to the constructed Schema.
+  /// \return The first failure encountered, if any.
+  Status WithSchemas(const std::vector<std::shared_ptr<Schema>>& schemas);
+
+  Status WithMetadata(const KeyValueMetadata& metadata);
```
Comment (pitrou) [inline cpp/src/arrow/type.h]: `AddMetadata`. Other `WithMetadata` methods return a _new_ instance with added metadata, leaving `this` intact.

File: cpp/src/arrow/type.cc [hunk 7]
```diff
@@ -354,18 +388,19 @@ std::unordered_multimap<std::string, int> CreateNameToIndexMap(
   return name_to_index;
 }
 
+template <int NotFoundValue = -1, int DuplicateFoundValue = -1>
```
Comment (pitrou) [inline cpp/src/arrow/type.cc]: Make it always `-1` and `-2` and use static constants to designate those values?

Comment (pitrou) [reply cpp/src/arrow/type.cc]: (can use `i < 0` to check that lookup failed)

File: cpp/src/arrow/type.cc [hunk 8]
```diff
@@ -716,6 +740,111 @@ std::vector<std::string> Schema::field_names() const {
...
+
+SchemaBuilder::SchemaBuilder(std::vector<std::shared_ptr<Field>> fields,
+                             ConflictPolicy policy)
+    : fields_(std::move(fields)),
+      name_to_index_(CreateNameToIndexMap(fields_)),
+      policy_(policy) {}
+
+SchemaBuilder::SchemaBuilder(const std::shared_ptr<Schema>& schema,
+                             ConflictPolicy conflict_resolution)
+    : fields_(std::move(schema->fields())),
```
Comment (pitrou) [inline cpp/src/arrow/type.cc]: No `std::move` here. `Schema::fields` returns a `const&` vector.

File: cpp/src/arrow/type.cc [hunk 9]
```diff
@@ -716,6 +740,111 @@ std::vector<std::string> Schema::field_names() const {
...
+SchemaBuilder::SchemaBuilder(ConflictPolicy policy) : policy_(policy) {}
+
+SchemaBuilder::SchemaBuilder(std::vector<std::shared_ptr<Field>> fields,
+                             ConflictPolicy policy)
+    : fields_(std::move(fields)),
+      name_to_index_(CreateNameToIndexMap(fields_)),
+      policy_(policy) {}
+
+SchemaBuilder::SchemaBuilder(const std::shared_ptr<Schema>& schema,
+                             ConflictPolicy conflict_resolution)
```
Comment (pitrou) [inline cpp/src/arrow/type.cc]: Why not `policy` as above?

**Review by pitrou — CHANGES_REQUESTED** [review]
Some comments below.

File: cpp/src/arrow/type.cc (hunk 7 above)
Comment (fsaintjacques) [reply cpp/src/arrow/type.cc]: This is how I initially did it, but it breaks the current API (external consumers of arrow). I can fix it internally, but our user will suffer.

File: cpp/src/arrow/type.cc (hunk 9 above)
Comment (fsaintjacques) [reply cpp/src/arrow/type.cc]: vim language-server-client didn't rename all of them it seems.

File: cpp/src/arrow/type.h (hunk 5 above)
Comment (fsaintjacques) [reply cpp/src/arrow/type.h]: Should I remove the plural and let C++ do the type matching on the parameter, e.g.

```c++
AddField(const std::shared_ptr<Field>& field);
AddField(const std::vector<std::shared_ptr<Field>>& fields);
```

Comment (pitrou) [reply cpp/src/arrow/type.h]: I prefer explicit plurals.

File: cpp/src/arrow/type_test.cc (hunk 4 above)
Comment (fsaintjacques) [reply cpp/src/arrow/type_test.cc]: Yes, in the future we can have a `struct MergeOptions`, e.g. if it should tolerate numeric promotion, etc.

File: cpp/src/arrow/type.h [hunk 10]
```diff
@@ -24,6 +24,7 @@
 #include <iosfwd>
 #include <memory>
 #include <string>
+#include <unordered_map>
```
[the remark this answers is not in the document]
Comment (fsaintjacques) [reply cpp/src/arrow/type.h]: What do you suggest, pimpl, or unique_ptr?

File: cpp/src/arrow/type.h (hunk 1 above)
Comment (fsaintjacques) [reply cpp/src/arrow/type.h]: Settled to `IGNORE` and `REPLACE`.

File: cpp/src/arrow/type.cc [hunk 11]
```diff
@@ -732,43 +872,22 @@ Result<std::shared_ptr<Schema>> UnifySchemas(
...
+    if (!schema->HasDistinctFieldNames()) {
+      return Status::Invalid("Can't unify schema with duplicate field names.");
     }
+    RETURN_NOT_OK(builder.AddSchema(schema));
   }
 
-  return schema(std::move(fields))->WithMetadata(schemas[0]->metadata());
-}
+  return builder.Finish();
+}  // namespace arrow
```
Comment (fsaintjacques) [inline cpp/src/arrow/type.cc]: urg, clang-format bug.

File: cpp/src/arrow/type.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/type.h]: Both?

File: cpp/src/arrow/type.cc (hunk 7 above)
Comment (pitrou) [reply cpp/src/arrow/type.cc]: Nevermind then, let's keep it like this.

**Comment by fsaintjacques** [comment]
@brills, please take a look.

**Review by bkietz — APPROVED** [review]
This is a nice foundation for interactive as well as programmatic schema reconciliation. Thanks for doing this!

**Pull request closed without merging**